enhancements/hide_container_mountpoints.md

```diff
@@ -0,0 +1,139 @@
+1819868](https://bugzilla.redhat.com/show_bug.cgi?id=1819868), but we can work
+around this exclusively within openshift. Using a separate mountpoint to for
+both crio-o and kubelet mounts can completely segregate all container-specific
+mounts away from any systemd interaction whatsoever.
+
+### Goals
+
+- cri-o places all container mountpoints in a separate mount namespace from systemd
+- kubelet places all container mountpoints in this same separate mount namespace from systemd
+- systemd does not see any cri-o or kubelet mountpoints
+- cri-o containers still see all appropriate kubelet mountpoints
+- Restarting either crio.service or kubelet.service does not result in the namespaces getting out-of-sync
+
+### Non-Goals
+
+- Fix systemd mountpoint scanning
+- Change cri-o pinns to support pinning mount namespaces
+
+## Proposal
+
+We can create a separate mount namespace and cause both crio and kubelet to
+launch within it to hide their many many mounts from systemd by creating:
+
+- A service called container-mount-namespace.service which spawns a separate
+  namespace (via systemd's 'PrivateMounts' mechanism), and then sleeps forever.
+  We don't want to create the namespace in crio.service or kubelet.service,
+  since if either one restarts they would lose eachother's namespaces.
+
+- An override file for each of crio.service and kubelet.service which wrap the
+  original command under 'nsenter' so they both use the mount namespace created
+  by 'container-mount-namespace.service'
+
+With these in place, both kubelet and crio create their mounts in the new
+shared (with eachother) but private (from systemd) namespace.
+
+### User Stories
+
+The end-user experience should not be affected in any way by this proposal, as
+there is no outward API changes. There is some supportability change though,
+since anyone attempting to inspect the cri-o or kubelet mountpoints externally
+would need to be aware that these are now available in a different namespace
+than the default top-level systemd mount namespace.
+
+### Implementation Details/Notes/Constraints
+
+Here is an example of the container-mount-namespace.service:
+
+    [Unit]
+    Description=Manages a mount namespace that both kubelet and crio can use to share their container-specific mounts
+    
+    [Service]
+    PrivateMounts=on
+    ExecStart=bash -c 'while :; do sleep infinity; done'
+
+This needs to be managed separately from either crio or kubelet to avoid the
+namespaces getting out of sync if either of those services restart. The systemd
+'PrivateMounts=on' facility does a nice job of creating a separate mountpoint,
+and both crio and kubelet can find the associated namespace via 'nsenter' as
+follows:
+
+    nsenter -m -t $(systemctl show --property MainPID --value container-mount-namespace.service) $ORIGINAL_EXECSTART
+
+### Risks and Mitigations
+
+Any other external or 3rd-party tools would need to change to match.
```

Comment:
a field should be added in open questions to investigate who may be affected by this

Reply:
Yeah big worry is about anything external that cares about the mounts.

Reply:
i thought we hit perf issue with these mounts and being inspected in systemd cadvisor handler.

https://github.com/google/cadvisor/pull/1573

i am not aware who actually cares about the mounts in question at this point...

Reply:
Podman/Skopeo/Buildah will also need to run in the same mount namespace, otherwise they will end-up setting up the storage in the host mount namespace, possibly causing corruption of the storage

Reply:
CSI driver pods / containers typically mount `/var/lib/kubelet` into their pods with `rshared` mount propagation. They must mount volumes there and kubelet + the container runtime must see them there. Any change in this behavior will likely break storage in OCP clusters.

What will a pod with `/var/lib/kubelet` HostPath volume actually see? The systemd or kubelet mount namespace?

Reply:
Good question. I'll double-check this!

Reply:
I hope I've done this right, but when I run a pod with a `/var/lib/kubelet` HostPath, the container is in its own mount namespace, but since this is inherited from the container runtime namespace (which is identical to the kubelet mount namespace), the pod does indeed see all the mounts under `/var/lib/kubelet`.

Do you have any pointers to a simple CSI pod I can play with to do some deeper testing of this?  Or is there a specific e2e suite I can run that specifically exercises the CSI architecture?

Reply:
@giuseppe : Can you give me a little more context or an example of how Podman/Skopeo/Buildah normally run in conjunction with OpenShift?  If they set up storage in the host mount namespace this will be propagated down to the container mount namespace (because it's a slave of the host mount namespace); so I'm unsure how much of a concern this is.

Reply:
I think a potential issue is with `/var/lib/containers/storage/overlay`, as it is always mounted `MS_PRIVATE` and it is not propagated to different namespaces.

Reply:
this idea is very similar to rootless `podman unshare` that executes the specified command in a different mount (and user in this case) namespace where all the containers are running.  For rootless it is a necessity because it needs a user namespace to gain capabilities.

I think in future we could teach other container tools to check if there are images/containers mounted in a different mount namespace and try to join it, or at least raise a warning.

Reply:
@giuseppe : When you mention `/var/lib/containers/storage/overlay`, I'd love to learn more about that filesystem and how it's used and shared between various tools, especially since it's mounted `MS_PRIVATE`.

If you don't mind taking the time to explain, I'd like to understand more about which process mounts the filesystem there, and which other processes expect to be able to act on it.

Reply:
any containers tool based on containers/storage (so all of CRI-O, Podman, Skopeo, Buildah).

The first access to the storage makes sure that `/var/lib/containers/storage/$DRIVER` is a mount point and it is made `MS_PRIVATE`.  All containers and images are mounted under this directory.  There is also some cleanup performed when there are no more containers and images mounted.

The reason for `MS_PRIVATE` is to prevent that each mountpoint will be propagated to the mount namespaces for each newly created container and avoid the cost of unmounting them.